### ISSUE TITLE
Consider exposing a rawer interface to array indexing

### ISSUE THREAD

**Issue by treeowl** [haskell/primitive#79]
Currently, the only interface to `indexArray#` in `primitive` is

```haskell
indexArrayM :: Monad m => Array a -> Int -> m a
```

This isn't the greatest when trying to use an array element to produce an unlifted result. Take the simplest thing: `ixArr# :: Array a -> Int -> (# a #)`. How can I implement this in terms of `indexArrayM`? Well, I need to pick a `Monad` instance. Let me dig through my bag of Monads:

1. `Identity`: not a chance.
2. `ST`: nope, that won't do it either.
3. `State`: gotta be kidding.
4. `Control.Monad.Trans.Cont.ContT`: Almost barely! It just needs to be rewritten to be levity polymorphic.
5. `Control.Monad.Codensity`: Almost barely! It just needs to be rewritten to be levity polymorphic.

Anyway, I've run out of monads I can pull off the shelf. So I can

1. Write my own version of `Cont` with the kind I want.
2. Implement the "lifted identity" monad `data Box a = Box a`.
3. Forget `primitive` and dig into `GHC.Prim`.

There are three simple ways to fix the problem:

1. `index# :: Monad m => Array a -> Int -> (# a #)`, boxing the array and `Int` but leaving the tuple bare.
2. Define the `Box` type and offer `indexBox :: Array a -> Int -> Box a`.
3. Offer a CPS version (unwrapping the custom `Cont`): `indexC :: forall a (r :: TYPE rep). Array a -> Int -> (a -> r) -> r`

**Comment by RyanGlScott** [comment]
Well, you now have a levity polymorphic `Codensity` in https://github.com/ekmett/kan-extensions/pull/52. Does this scratch your itch sufficiently?

**Comment by treeowl** [comment]
Well... it's an option now, but is that really the answer?

**Comment by RyanGlScott** [comment]
Well, you tell me! As you've noted, there's not a clear single way to work around this problem in `primitive` itself, as one could define at least three variants of `indexArrayM` that allow one to work over unlifted types. Given the uncertainty here, and the fact that any of these variations could be defined with relative ease, perhaps we should just endorse that as the "solution"?

**Comment by treeowl** [comment]
I actually think my CPS variant is the right solution. Until we get
levity-polymorphic newtypes, both (modified) ContT and Codensity can only
work for this nicely by the skin of their teeth, because (# a #) is the
application of a constructor to a value. Suppose I want to produce an Int#
instead; now I probably need to first get (# a #) and then match on it.
What I really want is kind-polymorphic Cont (as opposed to ContT), and
that's not yet available. But when you look at the implementation, you
quickly see that using pure is just a convenience for Applicative
situations; what we're really doing is taking a continuation (wrapped in a
dictionary). All I'm suggesting is that we offer an explicit version.

**Comment by RyanGlScott** [comment]
OK, that's a more definitive statement than before!

I think if you whipped up a patch offering this, it'd likely get accepted.

**Comment by treeowl** [comment]
Sorry, I mean kind-polymorphic Cont is not yet available off the shelf and
can't be built really nicely from kind-polymorphic ContT; strange hoops
must be jumped through.

**Comment by treeowl** [comment]
Having worked with this a little more, I believe we should at least offer

```haskell
indexWhatever# :: Array a -> Int -> (# a #)
```

because its type makes its purpose clear without any other context. The CPS version is nice and Haskell2010-looking, but the ergonomics probably aren't great. The `Box` approach is fine, and arguably more consistent with the rest of the interface, but people will have to look up what a `Box` is. Perhaps we should offer both `indexWhatever#` and `indexWhateverBox`?

**Comment by RyanGlScott** [comment]
I would be perfectly happy with either an unboxed tuple version or a `Box` version, since I believe that regardless of the choice, we _should_ document what use cases it has. At least, it's not quite as obvious for me what its purpose is as it is for you :)

**Comment by treeowl** [comment]
@RyanGlScott, let me approach this from another angle; maybe you'll be able to convert what I say here into a better explanation. Suppose I have a type

```haskell
data Array3 a = Array3 { a0 :: a, a1 :: a, a2 :: a }
```

There are two ways to get elements from an `Array3 a`: pattern matching and record selectors. I can write something like

```haskell
app2 :: (a -> b) -> Array3 a -> b
app2 f = f . a2
```

That's a perfectly valid function, lazy in its second argument:

```haskell
app2 (const 3) undefined = 3
```

This laziness can potentially cause performance trouble. Suppose I write

```haskell
g a b = if b then a else 3
```

If I call `app2 g arr`, I get a function that takes a `Bool` and depending on its value either uses the result of the lookup or doesn't. This means GHC may need to produce a thunk to perform the lookup (a performance problem of its own) and that thunk may keep the whole `Array3` live. (The selector thunk optimization should help with the latter here, but it won't with a real array.)

In the common case where we don't want this laziness, the most direct thing is to pattern match:

```haskell
app2' :: (a -> b) -> Array3 a -> b
app2' f (Array3 _ _ x) = f x
```

Now we perform the "lookup" first, and then call `f` with the result. The potential leak is gone, in exchange for being strict in the `Array3`.

In this context we have another option: force the `Array3` and let GHC figure out the rest:

```haskell
app2'' :: (a -> b) -> Array3 a -> b
app2'' f !arr= f (a2 arr)
```

Now let's get to real arrays. `flip indexArray 2` is very similar to the `a2` accessor above, and lets us write the lazy version. Unfortunately, `app2''` isn't an option: if we write

```haskell
app2Urp :: (a -> b) -> Array a -> b
app2Urp f !arr = f (indexArray arr 2)
```

then we force the array, but because GHC doesn't know the index is valid, it can't actually perform the lookup! Pattern matching isn't an option either, because we can't pattern match into an array. Hmph. For a solution, imagine a variation on `Array3`:

```haskell
data Box a = Box a
data Array3B a = Array3B {ab0 :: !(Box a), ab1 :: !(Box a), ab2 :: !(Box a)}
```

Now we can write

```haskell
app2b :: (a -> b) -> Array3B a -> b
app2b f arr = case ab2 arr of
                           Box x -> f x
```

So one way to think of it is that `Array` is not an array that lazily holds elements, but rather one that strictly holds `Box`es. But the primops don't know anything about `Box`; they instead have `(# _ #)`.

**Comment by treeowl** [comment]
@RyanGlScott , did you get a chance to read my last comment? Did that help? Would it be worth adding something similar (but shorter) to the Haddocks in `primops.txt.pp`?

**Comment by RyanGlScott** [comment]
Your comment makes sense. I was only requesting that you put some subset of this wisdom into the Haddocks/comments for the benefit of others.

**Comment by treeowl** [comment]
@RyanGlScott, do you think you can help me formulate a shorter version that will still make sense?

**Comment by RyanGlScott** [comment]
I doubt I'm qualified to do so, but the part of about strictly holding boxes seems to be the key bit that would make someone choose `indexWhatever#` over `indexArrayM`, yes? Or perhaps I'm butchering the finer details, but that's the sort of thing I'd want to see in the Haddocks for `indexWhatever#`.

**Comment by treeowl** [comment]
`indexArrayM` is fine. It's just annoying to use in a pure context. `indexArray` is the one that tends to be problematic.

**Comment by treeowl** [comment]
Closed in #143.
